db/db_impl/db_impl_open.cc

```diff
@@ -1929,4 +1929,330 @@ Status DBImpl::Open(const DBOptions& db_options, const std::string& dbname,
+    }
+    this_cfd->mem()->SetNextLogNumber(max_log_number);
+  }
+  for (size_t i = 0; i < all_db_impls.size(); i++) {
+    auto* db = all_db_impls[i];
+    bool check_log_number = (i == 0 || merge_options.allow_source_write) &&
+                            merge_options.merge_memtable;
+    if (s.ok() && check_log_number && max_log_number != db->logfile_number_) {
+      assert(max_log_number > db->logfile_number_);
+      // Create a new WAL so that future memtable uses the correct log
+      // number.
+      log::Writer* new_log = nullptr;
+      s = db->CreateWAL(max_log_number, 0 /*recycle_log_number*/,
+                        0 /*preallocate_block_size*/, &new_log);
+      if (s.ok()) {
+        db->logfile_number_ = max_log_number;
+        assert(new_log != nullptr);
+        db->logs_.emplace_back(max_log_number, new_log);
+        auto current = db->versions_->current_next_file_number();
+        if (current <= max_log_number) {
+          db->versions_->FetchAddFileNumber(max_log_number - current + 1);
+        }
+      }
+    }
+    db->log_write_mutex_.Unlock();
+    db->write_thread_.ExitUnbatched(all_writers[i].get());
+    db->mutex_.Unlock();
+  }
+  if (!s.ok()) {
+    return s;
+  }
+
+  TEST_SYNC_POINT("DBImpl::MergeDisjointInstances:AfterMergeMemtable:1");
+
+  // Merge table files.
+  assert(s.ok());
+  autovector<VersionEdit> cf_edits;
+  auto* this_table_cache = table_cache_.get();
+  for (size_t cf_i = 0; cf_i < this_cfds.size(); cf_i++) {
+    auto* this_cfd = this_cfds[cf_i];
+    auto* this_table_factory = this_cfd->ioptions()->table_factory.get();
+    cf_edits.emplace_back();
+    cf_edits.back().SetColumnFamily(this_cfd->GetID());
+    auto& edit = cf_edits[cf_i];
+    for (size_t db_i = 0; db_i < db_impls.size(); db_i++) {
+      auto* super_version = cf_super_versions[cf_i][db_i];
+      if (super_version == nullptr) {
+        continue;
+      }
+      auto* db = db_impls[db_i];
+      auto* table_cache = db->table_cache_.get();
+      VersionStorageInfo& vsi = *super_version->current->storage_info();
+      auto& cf_paths = super_version->cfd->ioptions()->cf_paths;
+      auto GetDir = [&](size_t path_id) {
+        // Matching `TableFileName()`.
+        if (path_id >= cf_paths.size()) {
+          assert(false);
+          return cf_paths.back().path;
+        } else {
+          return cf_paths[path_id].path;
+        }
+      };
+      for (int level = 0; level < vsi.num_levels(); ++level) {
+        const auto& level_files = vsi.LevelFiles(level);
+        for (const auto& f : level_files) {
+          assert(f);
+          uint64_t target_file_number = versions_->FetchAddFileNumber(1);
+          std::string src =
+              MakeTableFileName(GetDir(f->fd.GetPathId()), f->fd.GetNumber());
+          std::string target = MakeTableFileName(
+              this_cfd->ioptions()->cf_paths.front().path, target_file_number);
+          s = GetEnv()->LinkFile(src, target);
```

Comment:
Can we introduce a `MergeHelper` to dispose something like link file, transfer table cache and checks. Make this function as a main workflow and hide the details.

Reply:
Which part exactly is making it difficult to read? It doesn't seem very wise to move out an integral part to something like `fn copy_file(source_db, source_cfd, target_db, target_cfd, level, source_fd, target_number, edit, ...)`

Reply:
The whole progress is clear, but the details exhaust me. However, I can't come up with a specified idea of how to refactor it.

Reply:
Added more comments. Not sure if it helps.